exercises/practice/matrix/.meta/test_template.tera

```diff
@@ -0,0 +1,18 @@
+use {{ crate_name }}::*;
+
+{% for test in cases %}
+#[test]
+{% if loop.index != 1 -%}
+#[ignore]
+{% endif -%}
+fn {{ test.description | slugify | replace(from="-", to="_") }}() {
+    let matrix = Matrix::new({{ test.input.string | json_encode() }});
+    assert_eq!(matrix.{{ test.property }}({{ test.input.index }}), Some(vec!{{ test.expected | json_encode() }}));
+    {% if loop.index == 4 -%}
+    assert_eq!(matrix.column({{ test.input.index }}), None);
+    {% endif -%}
+    {% if loop.index == 7 -%}
+    assert_eq!(matrix.row({{ test.input.index }}), None);
+    {% endif -%} 
```

Comment:
This is well intentioned, but we should stick to problem-specifications as closely as possible. If we find that some test cases are missing, it is better to add them upstream. (In rare cases where Rust-specific tests make sense, there can be an `additional-tests.json` file that is merged with the standard test cases.)

I don't think we need to check explicitly that this returns `None`, what else could it possibly do? Maybe an empty vector?

I am surprised upstream has not a single error handling test case. maybe that it's worth upstreaming at least one.

Reply:
I did find it strange as well, it really feels like there should be some kind of error handling test. Most other exercises assume at least one test is going to have invalid input. 

However, pushing it upstream would cause breaking changes in other repos I imagine? Obviously I'm new here so if you think it's a better option to just switch it to an empty vec, or keep it the same and not specifically test for a `None` output, just let me know.

Reply:
> However, pushing it upstream would cause breaking changes in other repos I imagine?

That's not a problem. Updating an exercise with new tests from upstream is generally a manual process per track (though my long-term goal with the generator is to automate much of that work for the Rust track). If some language track wants to update and absolutely does not want to do error handling, a language track can always exclude that specific test in `.meta/tests.toml`.

Would you mind opening a forum post about such a new test case and presumably follow that up with a PR?

I think the `Option<Vec<_>>` return type is definitely idiomatic, let's keep it. We can either wait for the error test to be upstreamed or delete these special cases to keep this PR unblocked. Up to you.

Reply:
> If some language track wants to update and absolutely does not want to do error handling, a language track can always exclude that specific test in .meta/tests.toml.

Totally forgot about this!

I would be happy to open the forum post and I'll wait to see what the feedback is before proceeding on this piece.

Reply:
I guess Isaac's argument makes sense to not have the input validation at all. The fact that Rust has such a focus on handling errors explicitly leaves us in a bit of an akward position though. `-> Option<Vec<_>>` is the idiomatic return type, I wouldn't change it. Otherwise we're forcing bad code on users. But it's weird to not have a single test case with `None`. I think two options make sense at this point, I'll leave the choice up to you:

- Not do any input validation at all, as intended by upstream.
- Add custom tests with `None` to `.meta/additional-tests.json`.

Reply:
I agree that he made valid points, and I definitely understand what he's saying. I'm personally still in the mindset of some kind of invalid input test makes sense for this particular exercise, but can see where it may cause larger issues.

I did opt to add in some additional tests -

- I added the `additional-tests.json` file
- I created two new UUIDs (hopefully that was what I was supposed to do)
- Added comments
- Regenerated

Let me know if that works.